### PR TITLE
Introduce a mechanism for services to be definable outside of the core binary

### PR DESCRIPTION
**Pull request by cramforce**
and to depend on them asynchronously. Use it for the Cid service that will be defined in amp-analytics.

Support for #871

@erwinmombay This will also be useful an an abstraction for the consent dialog.

### CODE REVIEW

File: src/service/cid-impl.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
+/**
```
Comment (cramforce) [inline src/service/cid-impl.js]: This is a copy, but GH diffs don't show it.

File: build-system/tasks/presubmit-checks.js [hunk 2]
```diff
@@ -48,6 +48,7 @@ var forbiddenTerms = {
     message: requiresReviewPrivacy,
     whitelist: [
       'src/cid.js',
+      'src/service/cid-impl.js',
```
Comment (erwinmombay) [inline build-system/tasks/presubmit-checks.js]: awesome, i was just thinking yesterday wanting to suggest we add a service folder :+1:

File: src/cid.js [hunk 3]
```diff
@@ -15,275 +15,13 @@
...
 /**
  * @param {!Window} window
- * @return {!Cid}
+ * @return {!Promise<!Cid>}
  */
 export function cidFor(window) {
-  return getService(window, 'cid', () => {
-    return new Cid(window);
-  });
+  return getElementService(window, 'cid', 'amp-analytics');
```
Comment (erwinmombay) [inline src/cid.js]: missing import of `getElementService`

Comment (cramforce) [inline src/cid.js]: Ick. Done.

File: src/service.js [hunk 4]
```diff
@@ -31,16 +44,85 @@ import {assert} from './asserts';
...
+ * Users should typically wrap this is a special purpose function (e.g.
+ * viewportFor(win)) for type safety and because the factory should not be
+ * passed around.
+ * @param {!Window} win
+ * @param {string} id of the service.
+ * @param {string} provideByElement Name of the custom element that provides
+ *     the implementation of this service.
+ * @return {!Promise<*>}
+ */
+export function getElementService(win, id, providedByElement) {
```
Comment (erwinmombay) [inline src/service.js]: Thanks!

File: src/service.js [hunk 5]
```diff
@@ -31,16 +44,85 @@ import {assert} from './asserts';
...
+    }
+  }
+  return s.obj;
+}
+
+/**
+ * Returns a promise for a service for the given id and window. Also expects
+ * an element that has the actual implementation. The promise resolves when
+ * the implementation loaded.
+ * Users should typically wrap this is a special purpose function (e.g.
```
Comment (erwinmombay) [inline src/service.js]: wrap this as?

File: src/service.js [hunk 6]
```diff
@@ -31,16 +44,85 @@ import {assert} from './asserts';
...
+      'Service %s was requested to be provided through %s, ' +
+      'but %s is not loaded in the current page.',
+      id, providedByElement, providedByElement);
+  const services = getServices(win, id);
+  const s = services[id];
+  if (s) {
+    if (s.promise) {
+      return s.promise;
+    }
+    if (s.obj) {
```
Comment (erwinmombay) [inline src/service.js]: can you explain the `s.obj` usage?

**Comment by erwinmombay** [comment]
LGTM, besides some minor questions and getting green tests.

File: src/service.js (hunk 5 above)
Comment (cramforce) [inline src/service.js]: Done

File: src/service.js (hunk 6 above)
Comment (cramforce) [inline src/service.js]: Done

**Pull request merged**